GG/GG/ZList.h

```diff
@@ -47,33 +51,67 @@ namespace GG {
     screen.  Windows in the z-list are kept in front-to-back order.  No
     windows may share the same z-value.  Add, Remove, MoveUp, and MoveDown all
     also add/remove/move all descendent windows.*/
-class GG_API ZList : public std::list<Wnd*>
+class GG_API ZList
 {
 public:
+    ~ZList();
+
+    /** RenderOrderIterable can be iterated in back to front render order. */
+    struct RenderOrderIterable {
+        RenderOrderIterable(const std::list<std::shared_ptr<Wnd>>& list) :
+            m_list(list.crbegin(), list.crend())
+        {}
+
+        std::vector<std::shared_ptr<Wnd>>::const_iterator begin()
+        { return m_list.cbegin(); }
+
+        std::vector<std::shared_ptr<Wnd>>::const_iterator end()
+        { return m_list.cend(); }
+
+        private:
+        const std::vector<std::shared_ptr<Wnd>> m_list;
+    };
+
     /** \name Accessors */ ///@{
+    /** Return a RenderOrderIterable in back to front render order. */
+    RenderOrderIterable RenderOrder() const;
+
     /** Returns pointer to the window under the point pt; constrains pick to
         modal if nonzero, and ignores \a ignore if nonzero. */
-    Wnd* Pick(const Pt& pt, Wnd* modal, const std::set<Wnd*>* ignore = nullptr) const;
+    std::shared_ptr<Wnd> Pick(const Pt& pt, const std::shared_ptr<Wnd>& modal, const std::set<Wnd*>* ignore = nullptr) const;
     //@}
 
     /** \name Mutators */ ///@{
     /** Add() places \a wnd in front of the list. */
-    void Add(Wnd* wnd);
+    void Add(std::shared_ptr<Wnd> wnd);
 
-    bool Remove(Wnd* wnd);   ///< Removes \a wnd from z-order.
+    /** Remove \p wnd from the z-ordered list. */
+    bool Remove(const std::shared_ptr<Wnd>& wnd);
+    bool Remove(const Wnd* const wnd);
 
     /** Moves \a wnd from its current position to the beginning of list;
         updates wnd's z-value. */
-    bool MoveUp(Wnd* wnd);
+    bool MoveUp(const std::shared_ptr<Wnd>& wnd);
+    bool MoveUp(const Wnd* const wnd);
 
     /** Moves \a wnd from its current position to the end of list; updates
         wnd's z-value. */
-    bool MoveDown(Wnd* wnd);
+    bool MoveDown(const std::shared_ptr<Wnd>& wnd);
+    bool MoveDown(const Wnd* const wnd);
     //@}
 
 private:
-    Wnd*     PickWithinWindow(const Pt& pt, Wnd* wnd, const std::set<Wnd*>* ignore) const; ///< Returns pointer to the window under the point pt; constrains pick to wnd and its decendents, and ignores \a ignore if nonzero.
+    using iterator = std::list<std::shared_ptr<Wnd>>::iterator;
+
     iterator FirstNonOnTop();              ///< Returns iterator to first window in list that is non-on-top (returns end() if none found).
+
+    /** Return the first candidate that satisfies \p predicate with a non none result along with
+        the result. Remove any non-existent Wnd examined. */
+    template <typename T>
+    boost::optional<std::pair<iterator, T>> Find(
+        const std::function<boost::optional<T> (const std::shared_ptr<Wnd>&)>& predicate) const;
```

Comment:
I meant the lines, and 112 in particular, is types and various brackets nested several deep. splitting it over a couple lines might make it easier to understand what's a function name, what's a type, and what's a parameter name.